---
header-includes:
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \pagenumbering{gobble}
geometry: margin=0.75in
output: pdf_document
params:
  projectname: "Project"
  thedate: !r Sys.Date()
  pathd: "/"
  infof: ".info"
title: "Alignment Summary Report for `r params$projectname`"
author: "Institute for Genome Sciences (IGS) Trancriptomics Analysis Team"
date: "`r params$thedate`"      
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
library(rmarkdown)
library(knitr)
library(reshape2)
library(grid)
library(gridExtra)
library(ggplot2)
library(png)
library(dplyr)
library(magick)
```


```{r table, echo=FALSE, fig.show='hide', message=FALSE, warning=FALSE}
 
###add legen contd. to all table parts
 
direc=unlist(params$pathd)
setwd(direc)
fn=paste0(direc,"/Summary.txt")
info<-unlist(params$infof)
newimg<-paste0(direc, "/AlignmentTables")
lns = readLines(fn, n=1)
vec<-unlist(strsplit(lns, split = "\t"))
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
chop_n<- nrow(mystring)
if(chop_n < 21)
{
mystring$Total.Reads<- prettyNum(mystring$Total.Reads,  big.mark = ",",  scientific=FALSE)
mystring$Total.Mapped.Reads <- prettyNum(mystring$Total.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
mystring$Uniquely.Mapped.Reads<- prettyNum(mystring$Uniquely.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
mystring$Total.Features<-prettyNum(mystring$Total.Features,  big.mark = ",",  scientific=FALSE)
mystring$Features.With.Coverage<-prettyNum(mystring$Features.With.Coverage,  big.mark = ",",  scientific=FALSE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
t1<-mystring[,1:6]
t2<-mystring[,c(1, 7:11)]
#mytheme2<-ttheme_default()
mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")),colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

t1g<-tableGrob(t1, theme=mytheme)
t2g<-tableGrob(t2, theme=mytheme)

h1 <- grobHeight(t1g)
w1 <- grobWidth(t1g)
title1 <- textGrob("Table 1. Alignment statistics for all samples listing the proportion of reads mapped for each sample", gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h1, vjust=0)
gt1 <- gTree(children=gList(t1g, title1))
#grid.draw(gt)

h2 <- grobHeight(t2g)
w2 <- grobWidth(t2g)
title2 <- textGrob("Table 2. Alignment statistics for all samples listing the proportion of reads mapping to different genomic components,\n the number of genomic features with read coverage, and the average RPKM for each sample", gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h2, 
                  vjust=0)
gt2 <- gTree(children=gList(t2g, title2))
  ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_1_table1.png"))
  ggsave(gt2, path = newimg, width = 8, height = 10, filename = paste0("t2_1_table2.png"))
#ggsave(t1g, width = 8, height = 10, path = newimg, filename = "t1g_table1.png")
#ggsave(t2g, width = 8, height = 10, path = newimg, filename = "t2g_table2.png")
}

if(chop_n>20)
{
df<-split(mystring, rep(1:ceiling(chop_n/20), each=20, length.out=chop_n))
for(i in 1: length(df))
{
  if(i==1)
  {
    df_hold<-data.frame(df[i], check.names = FALSE)
  colnames(df_hold)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads",
                        "Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic", "Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
  df_hold$Total.Reads<- prettyNum(df_hold$Total.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Mapped.Reads <- prettyNum(df_hold$Total.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Uniquely.Mapped.Reads<- prettyNum(df_hold$Uniquely.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Features<-prettyNum(df_hold$Total.Features,  big.mark = ",",  scientific=FALSE)
  df_hold$Features.With.Coverage<-prettyNum(df_hold$Features.With.Coverage,  big.mark = ",",  scientific=FALSE)
  
  t1<-df_hold[,1:6]
  t2<-df_hold[,c(1, 7:12)]
  mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")), colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

  t1g<-tableGrob(t1, theme=mytheme)
  t2g<-tableGrob(t2, theme=mytheme)
  h1 <- grobHeight(t1g)
  w1 <- grobWidth(t1g)
  title1 <- textGrob("Table 1. Alignment statistics for all samples listing the proportion of reads mapped for each sample", gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h1, vjust=0)
  gt1 <- gTree(children=gList(t1g, title1))
  ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_",i,"_table1.png"))
  
  h2 <- grobHeight(t2g)
  w2 <- grobWidth(t2g)
  title2 <- textGrob("Table 2. Alignment statistics for all samples listing the proportion of reads mapping to different genomic components,\n the number of genomic features with read coverage, and the average RPKM for each sample", gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h2, vjust=0)
  gt2 <- gTree(children=gList(t2g, title2))
  ggsave(gt2, path = newimg, width = 8, height = 10, filename = paste0("t2_",i,"_table2.png"))
  }
  else
  {
  df_hold<-data.frame(df[i], check.names = FALSE)
  colnames(df_hold)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads",
                        "Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic",                     "Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
  df_hold$Total.Reads<- prettyNum(df_hold$Total.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Mapped.Reads <- prettyNum(df_hold$Total.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Uniquely.Mapped.Reads<- prettyNum(df_hold$Uniquely.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Features<-prettyNum(df_hold$Total.Features,  big.mark = ",",  scientific=FALSE)
  df_hold$Features.With.Coverage<-prettyNum(df_hold$Features.With.Coverage,  big.mark = ",",  scientific=FALSE)
  
  t1<-df_hold[,1:6]
  t2<-df_hold[,c(1, 7:12)]
  mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")), colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

  t1g<-tableGrob(t1, theme=mytheme)
  t2g<-tableGrob(t2, theme=mytheme)
  h1 <- grobHeight(t1g)
  w1 <- grobWidth(t1g)
  title1 <- textGrob("Table 1. Alignment statistics for all samples listing the proportion of reads mapped for each sample (CONTD.)", gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h1, vjust=0)
  gt1 <- gTree(children=gList(t1g, title1))
  ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_",i,"_table1.png"))
  
  h2 <- grobHeight(t2g)
  w2 <- grobWidth(t2g)
  title2 <- textGrob("Table 2. Alignment statistics for all samples listing the proportion of reads mapping to different genomic components,\n the number of genomic features with read coverage, and the average RPKM for each sample (CONTD.)", gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h2, vjust=0)
  gt2 <- gTree(children=gList(t2g, title2))
  ggsave(gt2, path = newimg, width = 8, height = 10, filename = paste0("t2_",i,"_table2.png"))
  }
}
}

```

```{r few, fig.show='asis', echo=FALSE,fig.width=10, fig.height=12, message=FALSE, fig.align='center'}
Tables_Split1<-list.files(newimg, pattern = "_table1.png", full.names = TRUE)
Tables_Split2<-list.files(newimg, pattern = "_table2.png", full.names = TRUE)
if(chop_n<10)
{
  grid.arrange(gt1, gt2, nrow=2, as.table = TRUE)
}
```

```{r disp1, fig.show='asis', echo=FALSE,fig.width=10, fig.height=12, message=FALSE, fig.align='center'}

if(chop_n>=10)
{
  include_graphics(Tables_Split1, dpi=250)
}


#t1I<-readPNG(paste0(newimg, "/t1g.png"))
#rastert1<-grid::rasterGrob(as.raster(t1I))
#grid.arrange(rastert1, top=textGrob("Table 1. Alignment statistics for all samples listing the proportion of reads mapped for each sample", gp=gpar(fontsize=9)))
```

```{r disp, fig.show='asis', echo=FALSE}
Tables_Split2<-list.files(newimg, pattern = "_table2.png", full.names = TRUE)
if(chop_n>=10)
{
  include_graphics(Tables_Split2, dpi=250)
}
#grid.arrange(t1g)
#t2I<-readPNG(paste0(newimg, "/t2g.png"))
#rastert2<-grid::rasterGrob(as.raster(t2I))
#grid.arrange(rastert2, top=textGrob("Table 2. Alignment statistics for all samples listing the proportion of reads mapping to different genomic components,\n the number of genomic features with read coverage, and the average RPKM for each sample", gp=gpar(fontsize=9)))
```
\blandscape
```{r mergeGrid, echo=FALSE, warning=FALSE, fig.height=8, fig.width=12}
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
frame<-data.frame(mystring$Sample_ID, mystring$Total.Mapped.Reads, mystring$Total.Reads, mystring$Percent.Mapped.Reads)
colnames(frame)<-c("SampleID", "Total.Mapped", "Total.Reads", "Percent.Mapped")

unmapp<-frame
unmapp$Unmapped<-(frame$Total.Reads-frame$Total.Mapped)
unmapped<-data.frame(unmapp$SampleID, unmapp$Total.Mapped, unmapp$Unmapped)
colnames(unmapped)<-c("SampleID", "Mapped", "Unmapped")
unmappedF<-melt(unmapped, id ="SampleID")
colnames(unmappedF)<-c("SampleID", "variable", "value")
plot2<-ggplot() + geom_bar(aes(y = value, x = SampleID, fill = rev(factor(unmappedF$variable))), 
                      data = unmappedF, stat="identity")+
  labs(x = "", y= "Reads", fill="", caption="")+theme(text = element_text(size=9), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom", axis.text=element_text(size=8))+ scale_fill_manual(values=c("#999999", "indianred1", "#56B4E9"),labels=c("Unmapped", "Mapped"))+ theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))

plot1<-frame %>%
  select(SampleID, Percent.Mapped) %>%
  na.omit() %>%
  ggplot() +
  geom_line(aes(x= frame$SampleID, y = frame$Percent.Mapped, group=1)) +
  ylab("% Mapped") +
  theme(axis.title.x = element_blank(), axis.text=element_text(size=12))+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+ 
  theme(panel.grid = element_blank())

grid.arrange(ggplotGrob(plot1), ggplotGrob(plot2), heights=c(0.75, 7), bottom=textGrob("Figure 1. Barplot illustrating the total reads, the proportion of reads mapped and the percent mapped for each sample",gp=gpar(fontsize=7)))
```
\elandscape

```{r barplots, echo=FALSE, fig.show='asis', warning=FALSE, fig.hold='asis'}
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
totalmapped<-data.frame(mystring$Sample_ID,mystring$Total.Mapped.Reads, mystring$Total.Reads)
totalmapped<-data.frame(mystring$Sample_ID,mystring$Total.Mapped.Reads, mystring$Total.Reads)
colnames(totalmapped)<-c("SampleID", "TotalMapped", "TotalReads")
totalbar<-melt(totalmapped, id ="SampleID")
#ggplot(totalbar, aes(x=totalbar$SampleID, y=totalbar$value, fill=factor(totalbar$variable)))+geom_bar(stat="identity",position="dodge")+coord_flip()+labs(x = "Samples", y= "", fill="",caption= "Figure 1. Barplot illustrating the total reads and the proportion of reads mapped for each sample")+theme(legend.position="top")+ theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))
```
```{r percent, echo=FALSE, warning=FALSE}
pmapped<-data.frame(mystring$Sample_ID,mystring$Percent.Mapped.Reads)
colnames(pmapped)<-c("SampleID", "PercentMappedReads")
#ggplot() + geom_bar(data = pmapped, aes(x=pmapped$SampleID, y=pmapped$PercentMappedReads), stat="identity", fill ="#FF1520")+coord_flip()+labs(x = "Samples", y= "Percent Mapped")+ scale_y_continuous(expand = c(0, 0))
```
\blandscape
```{r stack, echo=FALSE, warning=FALSE, fig.height=8, fig.width=12}
mta<-data.frame( mystring$Sample_ID, mystring$Percent.Genic, mystring$Percent.Intergenic)
colnames(mta)<-c("Sample_ID", "Percent.Genic", "Percent.Intergenic")
msta<-melt(mta, id ="Sample_ID")
#ggplot() + geom_bar(aes(y = value, x = Sample_ID, fill = rev(factor(msta$variable))), data = msta, stat="identity")+labs(x = "", y= "Percentages", fill="", caption="Figure 2. Stacked barplot illustrating the proportion of reads mapping to different genomic components for each sample")+theme(text = element_text(size=9), axis.text.x = element_text(angle=90, hjust=1), legend.position="top")+ scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"),labels=c("Percent Intergenic", "Percent Intronic", "Percent Exonic"))+ theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))

percentplot<-ggplot() + geom_bar(aes(y = value, x = Sample_ID, fill = rev(factor(msta$variable))), 
                    data = msta, stat="identity")+labs(x = "", y= "Percentages", fill="")+
  theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+
  scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"),
                    labels=c("Percent Intergenic", "Percent Genic"))+ 
  theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))
#ggsave(percentplot, path = newimg, filename = "percentplot.png")
#+ facet_wrap(~value, shrink = TRUE)

grid.arrange(percentplot, bottom=textGrob("Figure 2. Stacked barplot illustrating the proportion of reads mapping to different genomic components for each sample", gp=gpar(fontsize=7)))
#grid.arrange(percentplot, bottom="Figure 2. Stacked barplot illustrating the proportion of reads mapping to different genomic components for each sample")

myinfo <- read.table(info, check.names=FALSE)
myinfo <- myinfo[order(myinfo$V1),]
```
\elandscape


```{r boxplots, echo=FALSE, warning=FALSE, fig.hold='asis'}
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
mystring <- mystring[order(mystring$Sample_ID),]
tbox<-data.frame(mystring$Sample_ID, myinfo$V2, mystring$Total.Mapped.Reads, mystring$Total.Reads)
colnames(tbox)<-c("Sample_ID", "Condition", "TotalMapped", "TotalReads")
ttbox<-melt(tbox, id=c("Sample_ID", "Condition"))
#ggplot(ttbox, aes(x = ttbox$Condition, y = value, fill = variable)) + geom_boxplot()+labs(x = "", y= "Reads",caption= "Figure 3. Boxplot illustrating the total number of reads and the number of reads mapped for each condition.")+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), dotsize = 0.3, binwidth = .5)+ scale_fill_discrete(name = "")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")

ttbox$variable<-factor(ttbox$variable,levels = rev(levels(ttbox$variable)),ordered = TRUE)
#conditionM<-ggplot(ttbox, aes(x = ttbox$Condition, y = value, fill =ttbox$variable)) + geom_boxplot()+labs(x = "", y= "Reads")+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), dotsize = 0.3, binwidth = .5)+ scale_fill_discrete(name = "")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")

conditionM<-ggplot(ttbox, aes(x = ttbox$Condition, y = value, fill =ttbox$variable)) + geom_boxplot()+labs(x = "", y= "Reads")+ scale_fill_discrete(name = "")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")

#conditionM<-ggplot(ttbox, aes(x = ttbox$Condition, y = value, fill = variable)) + geom_boxplot()+labs(x = "", y= "Reads")+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), dotsize = 0.3, binwidth = .5)+ scale_fill_discrete(name = "")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")

grid.arrange(conditionM, bottom=textGrob("Figure 3. Boxplot illustrating the total number of reads and the number of reads mapped for each condition.", gp=gpar(fontsize=7)))

#grid.arrange(ggplot(ttbox, aes(x = ttbox$Condition, y = value, fill = variable)) + geom_boxplot()+labs(x = "", y= "Reads") + geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), dotsize = 0.3, binwidth = .5)+ scale_fill_discrete(name = "")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="top"), textGrob("Figure 3. Boxplot illustrating the total number of reads and the number of reads mapped for each condition."))

```

```{r proportion, echo=FALSE, warning=FALSE}
pmapped<-mystring$Percent.Mapped.Reads
percentmapped<-data.frame(myinfo[,1], myinfo[,2], pmapped)
#ggplot(stack(percentmapped), aes(x= percentmapped$myinfo...2., y= percentmapped$pmapped))+geom_boxplot( color="darkred")+labs(x = "", y= "Percent Mapped Reads", caption="Figure 4. Boxplot illustrating the proportion of reads mapped for each condition.")+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), dotsize = 0.3)+theme( axis.text.x = element_text(angle=90, hjust=1))

pmappedC<-ggplot(stack(percentmapped), aes(x= percentmapped$myinfo...2., y= percentmapped$pmapped))+geom_boxplot( color="darkred")+labs(x = "", y= "Percent Mapped Reads")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1))

#pmappedC<-ggplot(stack(percentmapped), aes(x= percentmapped$myinfo...2., y= percentmapped$pmapped))+geom_boxplot( color="darkred")+labs(x = "", y= "Percent Mapped Reads")+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), dotsize = 0.3)+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1))

grid.arrange(pmappedC, bottom=textGrob("Figure 4. Boxplot illustrating the proportion of reads mapped for each condition.", gp=gpar(fontsize=7)))

mbox<-data.frame(mystring$Sample_ID, myinfo[,2], mystring$Percent.Genic, mystring$Percent.Intergenic)
colnames(mbox)<-c("Sample_ID", "Condition", "Percent.Genic", "Percent.Intergenic")
mmbox<-melt(mbox)
```
\newline
\newline
```{r prop, echo=FALSE, warning=FALSE}
#ggplot(mmbox, aes(x = mmbox$Condition, y = value, color=variable)) + geom_boxplot()+labs(x = "", y= "Percentage Value", caption="Figure 5. Boxplot illustrating the proportion of reads mapping to different genomic components for each  condition.")+scale_color_manual(values=c("#56B4E9", "#E69F00","#999999"))+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), binwidth=3, dotsize = 0.3)+ scale_fill_discrete(name ="")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+theme(legend.title=element_blank())

propotionC<-ggplot(mmbox, aes(x = mmbox$Condition, y = value, color=variable)) + geom_boxplot()+labs(x = "", y= "Percentage Value")+scale_color_manual(values=c("#56B4E9", "#E69F00","#999999"))+ scale_fill_discrete(name ="")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+theme(legend.title=element_blank())

#propotionC<-ggplot(mmbox, aes(x = mmbox$Condition, y = value, color=variable)) + geom_boxplot()+labs(x = "", y= "Percentage Value")+scale_color_manual(values=c("#56B4E9", "#E69F00","#999999"))+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), binwidth=3, dotsize = 0.3)+ scale_fill_discrete(name ="")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+theme(legend.title=element_blank())
grid.arrange(propotionC, bottom=textGrob("Figure 5. Boxplot illustrating the proportion of reads mapping to different genomic components for each  condition.", gp=gpar(fontsize=7)))

#ggplot(mmbox, aes(x = mmbox$Condition, y = value, color=variable)) + geom_boxplot()+labs(x = "", y= "Percentage Value")+scale_color_manual(values=c("#56B4E9", "#E69F00","#999999"))+ geom_dotplot(binaxis='y', stackdir='center',position=position_dodge(1), binwidth=3, dotsize = 0.3)+ scale_fill_discrete(name ="")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+theme(legend.title=element_blank())+ ggtitle("Figure 5. Boxplot illustrating the proportion of reads mapping to different genomic components for each  condition.")
```
