---
header-includes:
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \pagenumbering{gobble}
geometry: margin=0.75in
output: pdf_document
params:
  projectname: "Project"
  thedate: !r Sys.Date()
  pathd: "/"
  infof: ".info"
title: "Alignment Summary Report for `r params$projectname`"
author: "Institute for Genome Sciences (IGS) Trancriptomics Analysis Team"
date: "`r params$thedate`"      
---

\  
\  
\  

The alignment report demonstrates how well the sequencing reads align to the reference sequence
and specifically to the exons in the reference annotation. Low percentage of mapped reads to
reference suggest that there might be contaminated reads coming from other sources instead of the
reference, or that there are reads of low quality present. Ideally, a high proportion of exonic
and a low proportion of intergenic regions are expected. High proportion of
intergenic alignment may indicate presence of DNA sequences. The alignments were generated by
HISAT2 for eukaryotes or by Bowtie2 for prokaryotes. Samtools was used to generate alignment
statistics. Table 1 and Figure 1 summarize how well the reads from the samples aligned to the
reference used. Proportions of genic and intergenic are reported (Table 2 and Figure 2).
The alignments statistics for each group are represented in Figures 3 and 4. The proportions
of genic and intergenic for each group are represented in the boxplot in Figure 5.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
library(rmarkdown)
library(knitr)
library(reshape2)
library(grid)
library(gridExtra)
library(ggplot2)
library(png)
library(dplyr)
#library(magick)
```


```{r table, echo=FALSE, fig.show='hide', message=FALSE, warning=FALSE}
 
###add legen contd. to all table parts
 
direc=unlist(params$pathd)
dir.create(file.path(direc, "/AlignmentFiles/"), showWarnings = FALSE)
setwd(direc)
fn=paste0(direc,"/Summary.txt")
info<-unlist(params$infof)
newimg<-paste0(direc, "/AlignmentFiles")
lns = readLines(fn, n=1)
vec<-unlist(strsplit(lns, split = "\t"))
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
chop_n<- nrow(mystring)
if(chop_n < 21)
{
mystring$Total.Reads<- prettyNum(mystring$Total.Reads,  big.mark = ",",  scientific=FALSE)
mystring$Total.Mapped.Reads <- prettyNum(mystring$Total.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
mystring$Uniquely.Mapped.Reads<- prettyNum(mystring$Uniquely.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
mystring$Total.Features<-prettyNum(mystring$Total.Features,  big.mark = ",",  scientific=FALSE)
mystring$Features.With.Coverage<-prettyNum(mystring$Features.With.Coverage,  big.mark = ",",  scientific=FALSE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
t1<-mystring[,1:6]
t2<-mystring[,c(1, 7:11)]
#mytheme2<-ttheme_default()
mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")),colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

t1g<-tableGrob(t1, theme=mytheme)
t2g<-tableGrob(t2, theme=mytheme)

h1 <- grobHeight(t1g)
w1 <- grobWidth(t1g)

title_wrap<-strwrap("Table 1. Alignment Summary: Statistics for all samples listing proportions of reads mapped for each sample. Low percent mapped can indicate either poor read quality or contamination with nucleic acids of another organism.", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

title1 <- textGrob(title_wrapped, gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h1, vjust=0)
gt1 <- gTree(children=gList(t1g, title1))
#grid.draw(gt)

h2 <- grobHeight(t2g)
w2 <- grobWidth(t2g)

title_wrap<-strwrap("Table 2. Alignment to genomic component:  Statistics for all samples listing the proportion of reads mapped to different genomic components, the number of genomic features with read coverage, and the average RPKM for each samples. High proportions of reads mapping to intergenic regions can indicate DNA contamination. Immature RNA contamination can cause a high proportion of the reads to map to intronic regions.", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

title2 <- textGrob(title_wrapped, gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h2, 
                  vjust=0)
gt2 <- gTree(children=gList(t2g, title2))
  ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_1_table1.png"))
  ggsave(gt2, path = newimg, width = 8, height = 10, filename = paste0("t2_1_table2.png"))
#ggsave(t1g, width = 8, height = 10, path = newimg, filename = "t1g_table1.png")
#ggsave(t2g, width = 8, height = 10, path = newimg, filename = "t2g_table2.png")
}

if(chop_n>20)
{
df<-split(mystring, rep(1:ceiling(chop_n/20), each=20, length.out=chop_n))
for(i in 1: length(df))
{
  if(i==1)
  {
    df_hold<-data.frame(df[i], check.names = FALSE)
  colnames(df_hold)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads",
                        "Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic", "Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
  df_hold$Total.Reads<- prettyNum(df_hold$Total.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Mapped.Reads <- prettyNum(df_hold$Total.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Uniquely.Mapped.Reads<- prettyNum(df_hold$Uniquely.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Features<-prettyNum(df_hold$Total.Features,  big.mark = ",",  scientific=FALSE)
  df_hold$Features.With.Coverage<-prettyNum(df_hold$Features.With.Coverage,  big.mark = ",",  scientific=FALSE)
  
  t1<-df_hold[,1:6]
  t2<-df_hold[,c(1, 7:11)]
  mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")), colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

  t1g<-tableGrob(t1, theme=mytheme)
  t2g<-tableGrob(t2, theme=mytheme)
  h1 <- grobHeight(t1g)
  w1 <- grobWidth(t1g)
  
  title_wrap<-strwrap("Table 1. Alignment Summary: Statistics for all samples listing proportions of reads mapped for each sample. Low percent mapped can indicate either poor read quality or contamination with nucleic acids of another organism.", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h1, vjust=0)
  gt1 <- gTree(children=gList(t1g, title1))
  ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_",i,"_table1.png"))
  
  h2 <- grobHeight(t2g)
  w2 <- grobWidth(t2g)
  
  title_wrap<-strwrap("Table 2. Alignment to genomic component:  Statistics for all samples listing the proportion of reads mapped to different genomic components, the number of genomic features with read coverage, and the average RPKM for each samples. High proportions of reads mapping to intergenic regions can indicate DNA contamination. Immature RNA contamination can cause a high proportion of the reads to map to intronic regions.", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

  title2 <- textGrob(title_wrapped, gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h2, vjust=0)
  gt2 <- gTree(children=gList(t2g, title2))
  ggsave(gt2, path = newimg, width = 8, height = 10, filename = paste0("t2_",i,"_table2.png"))
  }
  else
  {
  df_hold<-data.frame(df[i], check.names = FALSE)
  colnames(df_hold)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads",
                        "Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic",                     "Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
  df_hold$Total.Reads<- prettyNum(df_hold$Total.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Mapped.Reads <- prettyNum(df_hold$Total.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Uniquely.Mapped.Reads<- prettyNum(df_hold$Uniquely.Mapped.Reads,  big.mark = ",",  scientific=FALSE)
  df_hold$Total.Features<-prettyNum(df_hold$Total.Features,  big.mark = ",",  scientific=FALSE)
  df_hold$Features.With.Coverage<-prettyNum(df_hold$Features.With.Coverage,  big.mark = ",",  scientific=FALSE)
  
  t1<-df_hold[,1:6]
  t2<-df_hold[,c(1, 7:11)]
  mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")), colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

  t1g<-tableGrob(t1, theme=mytheme)
  t2g<-tableGrob(t2, theme=mytheme)
  h1 <- grobHeight(t1g)
  w1 <- grobWidth(t1g)
  title_wrap<-strwrap("Table 1. Alignment Summary: Statistics for all samples listing proportions of reads mapped for each sample. Low percent mapped can indicate either poor read quality or contamination with nucleic acids of another organism (CONTD).", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h1, vjust=0)
  gt1 <- gTree(children=gList(t1g, title1))
  ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_",i,"_table1.png"))
  
  h2 <- grobHeight(t2g)
  w2 <- grobWidth(t2g)
  title_wrap<-strwrap("Table 2. Alignment to genomic component:  Statistics for all samples listing the proportion of reads mapped to different genomic components, the number of genomic features with read coverage, and the average RPKM for each samples. High proportions of reads mapping to intergenic regions can indicate DNA contamination. Immature RNA contamination can cause a high proportion of the reads to map to intronic regions (CONTD).", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

  title2 <- textGrob(title_wrapped, gp=gpar(fontsize=7), y=unit(0.5,"npc")+ 0.75*h2, vjust=0)
  gt2 <- gTree(children=gList(t2g, title2))
  ggsave(gt2, path = newimg, width = 8, height = 10, filename = paste0("t2_",i,"_table2.png"))
  }
}
}

```

```{r few, fig.show='asis', echo=FALSE,fig.width=10, fig.height=12, message=FALSE, fig.align='center'}
Tables_Split1<-list.files(newimg, pattern = "_table1.png", full.names = TRUE)
Tables_Split2<-list.files(newimg, pattern = "_table2.png", full.names = TRUE)
if(chop_n<10)
{
  grid.arrange(gt1, gt2, nrow=2, as.table = TRUE)
}
```

```{r disp1, fig.show='asis', echo=FALSE, message=FALSE}

if(chop_n>=10)
{
  include_graphics(Tables_Split1, dpi=250)
}

```

```{r disp, fig.show='asis', echo=FALSE}
Tables_Split2<-list.files(newimg, pattern = "_table2.png", full.names = TRUE)
if(chop_n>=10)
{
  include_graphics(Tables_Split2, dpi=250)
}
```
#\blandscape
```{r mergeGrid, echo=FALSE, warning=FALSE, fig.height=8, fig.width=12}

mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
frame<-data.frame(mystring$Sample_ID, mystring$Total.Mapped.Reads, mystring$Total.Reads, mystring$Percent.Mapped.Reads)
colnames(frame)<-c("SampleID", "Total.Mapped", "Total.Reads", "Percent.Mapped")

unmapp<-frame
unmapp$Unmapped<-(frame$Total.Reads-frame$Total.Mapped)
unmapped<-data.frame(unmapp$SampleID, unmapp$Total.Mapped, unmapp$Unmapped)
colnames(unmapped)<-c("SampleID", "Mapped", "Unmapped")
unmappedF<-melt(unmapped, id ="SampleID")
colnames(unmappedF)<-c("SampleID", "variable", "value")
plot2<-ggplot() + geom_bar(aes(y = value, x = SampleID, fill = rev(factor(unmappedF$variable))), 
                      data = unmappedF, stat="identity")+
  labs(x = "", y= "Reads", fill="", caption="")+theme(text = element_text(size=9), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom", axis.text=element_text(size=8))+ scale_fill_manual(values=c("#999999", "indianred1", "#56B4E9"),labels=c("Unmapped", "Mapped"))+ theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))

plot1<-frame %>%
  select(SampleID, Percent.Mapped) %>%
  na.omit() %>%
  ggplot() +
  geom_line(aes(x= frame$SampleID, y = frame$Percent.Mapped, group=1)) +
  ylab("% Mapped") +
  theme(axis.title.x = element_blank(), axis.text=element_text(size=12))+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+ 
  theme(panel.grid = element_blank())

title_wrap<-strwrap("Figure 1. Alignment summary barplot: This illustrates the total reads, total mapped reads, and percent mapped for each sample. Low percent mapped can indicate either poor read quality or contamination with nucleic acids of another organism. Great differences in total mapped reads can cause biases in features detected.", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

hold_plot1 = grid.arrange(ggplotGrob(plot1), ggplotGrob(plot2), heights=c(0.75, 7), bottom=textGrob(title_wrapped,gp=gpar(fontsize=7)))
ggsave("Figure1.png", path = newimg, plot = hold_plot1, height = 8, width = 12)

```
\elandscape

```{r barplots, echo=FALSE, fig.show='asis', warning=FALSE, fig.hold='asis'}
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
totalmapped<-data.frame(mystring$Sample_ID,mystring$Total.Mapped.Reads, mystring$Total.Reads)
totalmapped<-data.frame(mystring$Sample_ID,mystring$Total.Mapped.Reads, mystring$Total.Reads)
colnames(totalmapped)<-c("SampleID", "TotalMapped", "TotalReads")
totalbar<-melt(totalmapped, id ="SampleID")

```
```{r percent, echo=FALSE, warning=FALSE}
pmapped<-data.frame(mystring$Sample_ID,mystring$Percent.Mapped.Reads)
colnames(pmapped)<-c("SampleID", "PercentMappedReads")

```
\blandscape
```{r stack, echo=FALSE, warning=FALSE, fig.height=8, fig.width=12}
mta<-data.frame( mystring$Sample_ID, mystring$Percent.Genic, mystring$Percent.Intergenic)
colnames(mta)<-c("Sample_ID", "Percent.Genic", "Percent.Intergenic")
msta<-melt(mta, id ="Sample_ID")
#ggplot() + geom_bar(aes(y = value, x = Sample_ID, fill = rev(factor(msta$variable))), data = msta, stat="identity")+labs(x = "", y= "Percentages", fill="", caption="Figure 2. Stacked barplot illustrating the proportion of reads mapping to different genomic components for each sample")+theme(text = element_text(size=9), axis.text.x = element_text(angle=90, hjust=1), legend.position="top")+ scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"),labels=c("Percent Intergenic", "Percent Intronic", "Percent Exonic"))+ theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))

percentplot<-ggplot() + geom_bar(aes(y = value, x = Sample_ID, fill = rev(factor(msta$variable))), 
                    data = msta, stat="identity")+labs(x = "", y= "Percentages", fill="")+
  theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+
  scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"),
                    labels=c("Percent Intergenic", "Percent Genic"))+ 
  theme(panel.grid = element_blank(), panel.border = element_blank())+ scale_y_continuous(expand = c(0, 0))
#ggsave(percentplot, path = newimg, filename = "percentplot.png")
#+ facet_wrap(~value, shrink = TRUE)

title_wrap<-strwrap("Figure 2. Alignment to Genomic Components: This illustrates the proportion of reads mapping to different genomics components. High proportions of reads mapping to intergenic regions can indicate DNA contamination. Immature RNA contamination can cause a high proportion of the reads to map to intronic regions.", width = 100, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

hold_plot2 = grid.arrange(percentplot, bottom=textGrob(title_wrapped, gp=gpar(fontsize=7)))
ggsave("Figure2.png", path = newimg, plot = hold_plot2, height = 8, width = 12)

myinfo <- read.table(info, check.names=FALSE)
myinfo <- myinfo[order(myinfo$V1),]
```
\elandscape


```{r boxplots, echo=FALSE, warning=FALSE, fig.hold='asis'}
mystring <- read.table(fn, check.names=FALSE, col.names=vec, sep="\t", as.is = TRUE)
colnames(mystring)<-c("Sample_ID","Total.Reads","Total.Mapped.Reads","Percent.Mapped.Reads","Percent.Properly.Paired","Uniquely.Mapped.Reads","Percent.Genic","Percent.Intergenic","Total.Features","Features.With.Coverage","Avg.RPKM" )
mystring <- mystring[order(mystring$Sample_ID),]
tbox<-data.frame(mystring$Sample_ID, myinfo$V2, mystring$Total.Mapped.Reads, mystring$Total.Reads)
colnames(tbox)<-c("Sample_ID", "Condition", "TotalMapped", "TotalReads")
ttbox<-melt(tbox, id=c("Sample_ID", "Condition"))


ttbox$variable<-factor(ttbox$variable,levels = rev(levels(ttbox$variable)),ordered = TRUE)

conditionM<-ggplot(ttbox, aes(x = ttbox$Condition, y = value, fill =ttbox$variable)) + geom_boxplot()+labs(x = "", y= "Reads")+ scale_fill_discrete(name = "")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")

title_wrap<-strwrap("Figure 3. Read Distribution by Group: The boxplot illustrates the total number of reads and the number of mapped reads for all samples in each group. This figure gives an indication of how well samples in each group mapped to the reference. The median line in each box indicates how even the distribution of total reads/ mapped reads is within the group.", width = 140, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

hold_plot3=grid.arrange(conditionM, bottom=textGrob(title_wrapped, gp=gpar(fontsize=7)))
ggsave("Figure3.png", path = newimg, plot = hold_plot3, height = 8, width = 12)

```

```{r proportion, echo=FALSE, warning=FALSE}
pmapped<-mystring$Percent.Mapped.Reads
percentmapped<-data.frame(myinfo[,1], myinfo[,2], pmapped)

#pmappedC<-ggplot(stack(percentmapped), aes(x= percentmapped$myinfo...2., y= percentmapped$pmapped))+geom_boxplot( color="darkred")+labs(x = "", y= "Percent Mapped Reads")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1))

pmappedC<-ggplot((percentmapped), aes(x= percentmapped$myinfo...2., y= percentmapped$pmapped))+geom_boxplot( color="darkred")+labs(x = "", y= "Percent Mapped Reads")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1))

title_wrap<-strwrap("Figure 4. Percent Mapped Reads: The boxplot illustrates the percentage of mapped reads for all samples in each group. This figure can highlight the differences in how reads map in different groups.", width = 140, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

#grid.arrange(pmappedC, top=textGrob(title_wrapped, gp=gpar(fontsize=7)))

hold_plot4 = grid.arrange(pmappedC, top=textGrob(title_wrapped, gp=gpar(fontsize=7)))
ggsave("Figure4.png", path = newimg, plot = hold_plot4, height = 8, width = 12)

mbox<-data.frame(mystring$Sample_ID, myinfo[,2], mystring$Percent.Genic, mystring$Percent.Intergenic)
colnames(mbox)<-c("Sample_ID", "Condition", "Percent.Genic", "Percent.Intergenic")
#mbox<-data.frame(mystring$Sample_ID, myinfo[,2], mystring$Percent.Exonic, mystring$Percent.Intronic, mystring$Percent.Intergenic)
#colnames(mbox)<-c("Sample_ID", "Condition", "Percent.Exonic", "Percent.Intronic", "Percent.Intergenic")
mmbox<-melt(mbox)
```

```{r prop, echo=FALSE, warning=FALSE}
propotionC<-ggplot(mmbox, aes(x = mmbox$Condition, y = value, color=variable)) + geom_boxplot()+labs(x = "", y= "Percentage Value")+scale_color_manual(values=c("#56B4E9", "#E69F00","#999999"))+ scale_fill_discrete(name ="")+theme(text = element_text(size=10), axis.text.x = element_text(angle=90, hjust=1), legend.position="bottom")+theme(legend.title=element_blank())

title_wrap<-strwrap("Figure 5. Genomic Components per condition: The boxplot illustrates the proportion of reads mapping to different genomic components for all samples in every group. The distristribution of reads between different components can be an indicator of alignment/reference quality or sequence quality.", width = 140, simplify = F)

title_wrapped<-sapply(title_wrap, paste, collapse = "\n")

hold_plot5 = grid.arrange(propotionC, bottom=textGrob(title_wrapped, gp=gpar(fontsize=7)))
ggsave("Figure5.png", path = newimg, plot = hold_plot5, height = 8, width = 12)

```
