---
header-includes:
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \pagenumbering{gobble}
- \DeclareUnicodeCharacter{2212}{-}
geometry: margin=0.75in
output: pdf_document
sansfont: Calibri Light
params:
  projectname: "Project"
  thedate: !r Sys.Date()
  pathd: "/"
  infof: ".info"
  mappingf: "none"
title: "DE Report for `r params$projectname`"
author: "Institute for Genome Sciences (IGS) Trancriptomics Analysis Team"
date: "`r params$thedate`"
---
\  
\  
\ 
The Differential Expression Report graphically demonstrates the differences between conditions or
groups. The raw counts generated by HTSeq are used as input for differential expression analysis
using either DESeq or DESeq2 Bioconductor packages. DESeq/DESeq2 is used to estimate gene
dispersion, normalize read count by library size to generate normalized counts for each gene, and
to statistically determine differentially expressed genes (DEGs) between two conditions/groups of
interest.

Table 1 summarizes the number of significantly up- and down-regulated genes for each comparison of
interest with significance defined using a False Discovery Rate (FDR) cutoff of FDR <= 0.05 and a
log2 fold change (LFC) >= +/- 1. For each comparison, a heatmap of the LFCs and gene expression for
the DEGs are displayed in order to show the size of the difference across all samples (Figure 1).
Figure 2 displays a MA plot for each comparison to visualize the relationship of the LFC in
expression of each gene between the two conditions against the log2 mean expression of the gene.

Additional plots are used to explore similarities and differences in DE between two or more
pairwise comparisons. These will only be generated if there exist more than 2 conditions. A Venn
diagram in Figure 3 demonstrates the relationships of significant DEGs among pairwise comparisons.
The final plot is a quadrant plot which is a scatter plot where each dot represents the LFC of the
expression of a gene in one comparison against the LFC of that gene's expression in another
comparison. Figure 4 displays a quadrant plot for each of the possible pairwise comparisons to
visualize whether or not the significant DEGs have the same or opposite relationship between two
comparisons.

\newpage
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rmarkdown)
library(knitr)
library(grid)
library(gridExtra)
library(gtools)
library(RColorBrewer)
library(gplots)
library(limma)
library(ggplot2)
library(ComplexHeatmap)
library(reshape2)
library(circlize)
library(ggfortify)
library(edgeR)
library(ggdendro)
library(ggpubr)
library(VennDiagram)
#library(magick)
library(png)
```

```{r DEfiles, echo=FALSE, include=FALSE}
direc=unlist(params$pathd)
mapf<-unlist(params$mappingf)
path=direc
setwd(direc)
infofile=unlist(params$infof)
DEpath<-paste0(path, "DE/")
filter_de_path=paste0(path,"/test_summary.txt")
DeFiles<-list.files(DEpath, pattern = ".txt", full.names = TRUE)
FDRc<-0.05
LFCc<-1
GEPathD=paste0(direc,"/GE_Outputs/")
DEOut<-paste0(path, "/DE_Outputs/")
dir.create(file.path(direc, "/DE_Outputs/"), showWarnings = FALSE)
all_counts_file<-paste0(direc, "/all_counts.txt")
```

```{r definitions, echo=FALSE, warning=FALSE}
match_IDs<-function(from_counts, from_info){
  if (! all(colnames(from_counts) == from_info$SampleID)) {
  cat(file=stderr(),
      'Error: The IDs of the matrix (on the first line) and the metadata id do not match.\n'
  )
    return(FALSE)
} else {
  cat(file=stderr(),
      'QC Checkpoint Passed.\n'
  )
  return(TRUE)
}
}
```

```{r table, echo=FALSE, warning=FALSE, fig.show='hide', message = FALSE}
DEpath<-paste0(path, "/DE/")
DECountPath<-paste0(path, "/DE/")
filtde_table=read.table(filter_de_path, check.names=FALSE, as.is = TRUE, header = T, blank.lines.skip = TRUE, sep="\t")
relfit_de=filtde_table[filtde_table$FDR==0.05 & startsWith(filtde_table$ReadCountPercentile,'0.1'),]
relfit_de$P.value=NULL
filtde_table=relfit_de[complete.cases(relfit_de),]
DeFiles<-list.files(DEpath, pattern = "de_genes.txt", full.names = TRUE)
table_DE<-data.frame()
comp_names<-(basename(DeFiles))
for(i in 1:length(comp_names))
{
  comp_names[i]<-gsub("\\..*","",comp_names[i])
}
table_DE<-data.frame(Comparision=comp_names)
table_DE$FDR<-FDRc
table_DE$LFC_UP<-LFCc
table_DE$LFC_DOWN<-(-LFCc)
up_reg<-vector()
down_regulated<-vector()
tot<-vector()

plot_list = list()
for(k in 1:length(DeFiles))
{
  comp_temp<-read.table(DeFiles[k], check.names=FALSE, as.is = TRUE, header = T, blank.lines.skip = TRUE)
  if("FC" %in% colnames(comp_temp)==FALSE)
  {
    comparision<-names(comp_temp)[5]
    names(comp_temp)[5]<-"LFC"
    holdFC<-logratio2foldchange(comp_temp$LFC)
    comp_temp$FC<-holdFC
    comp_temp<-comp_temp[, c(1,2,3,4,10,5,8,9)]
    fname<-basename(DeFiles[i])
    write.table(comp_temp, file = paste0(DEpath,fname), sep = "\t")
  }
  if("FC" %in% colnames(comp_temp)==TRUE)
  {
    comparision<-names(comp_temp)[6]
    names(comp_temp)[6]<-"LFC"
  }
  testcomp<-comp_temp[comp_temp$FDR<=FDRc, ]
  Fcomp<-testcomp[testcomp$LFC >=LFCc | testcomp$LFC<= -LFCc, ]
  DEInf<-Fcomp[order(Fcomp$Read.Count.All), ]
  ###changed from Fcomp to comp temp    
  p10comp<-(nrow(comp_temp)*10)/100
      p10compdf<-comp_temp[0:p10comp,]
      cutoff<-floor(mean(range(p10compdf$Read.Count.All)))
      ##rounded value instead of raw for counts
      Fcomp<-DEInf[round(DEInf$Read.Count.All) >= cutoff, ]
  up<-sum(Fcomp$LFC >0)
  down<-sum(Fcomp$LFC <0)
  total<-up+down
  up_reg[k]<-up
  down_regulated[k]<-down
  tot[k]<-total
}



table_DE$ReadCountPercentile<-0.1
table_DE$Up<-up_reg
table_DE$Down<-down_regulated
table_DE$Total<-tot

mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.5, fontface="plain")), colhead = list(fg_params=list(cex = 0.5, fontface="bold")),rowhead = list(fg_params=list(cex = 0.5, fontface="bold")))

#De_table_final<-tableGrob(table_DE, theme=mytheme)
De_table_final<-tableGrob(filtde_table, theme=mytheme)
save_table<-grid.arrange(De_table_final, top=textGrob("Summary of differentially expressed genes for each comparison based on filter cutoffs of FDR <= 0.05 and LFC >= +/- 1", gp=gpar(fontsize=9)))

h2 <- grobHeight(De_table_final)
w2 <- grobWidth(De_table_final)

title_wrap=strwrap("Table 1. Differentially Expressed Genes Summary: The numbers of of genes that are differentially expressed for each comparison base on filter cutoffs of false discovery rate (FDR) <= 0.05 and log fold change (LFC) >= +/- 1.", width=90, simplify = F)

title_wrapped=sapply(title_wrap, paste, collapse = "\n")

title_tree <- textGrob(title_wrapped, gp=gpar(fontsize=9), y=unit(0.5,"npc")+ 0.75*h2, vjust=0)

gtree1 <- gTree(children=gList(De_table_final, title_tree))

#ggsave(gt1, path = newimg, width = 8, height = 10, filename = paste0("t1_1_table1.png"))
  
  
#ggsave(gtree1, width = 8, height = 10, path =DEOut , filename = paste0("DE_table.png"))
```
\newpage
```{r disp1, fig.show='asis', echo=FALSE, warning=FALSE, message = FALSE}
DE_Table_img=list.files(DEOut, pattern = "DE_table.png", full.names = TRUE)
include_graphics(DE_Table_img, dpi=125)
```

```{r DEmaplots, echo=FALSE, message = FALSE}
for(i in 1: length(DeFiles))
{
DEInf<-read.table(DeFiles[i], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  if("FC" %in% colnames(DEInf)==FALSE)
  {
    comparision<-names(DEInf)[5]
    names(DEInf)[5]<-"LFC"
    holdFC<-logratio2foldchange(DEInf$LFC)
    DEInf$FC<-holdFC
    DEInf<-DEInf[, c(1,2,3,4,10,5,8,9)]
    fname<-basename(DeFiles[i])
    write.table(DEInf, file = paste0(DEpath,fname), sep = "\t")
  }
  if("FC" %in% colnames(DEInf)==TRUE)
  {
    comparision<-names(DEInf)[6]
    names(DEInf)[6]<-"LFC"
  }
#comparision<-names(DEInf)[6]
#names(DEInf)[6]<-"LFC"
columnI <- DEInf$LFC[which(!is.na(DEInf$LFC))]
columni <- columnI[which(columnI < Inf)]
posINF<-max(columni, na.rm = T)
columni <- columnI[which(columnI > -Inf)]
negINF<-min(columni, na.rm = T)
DEInf$LFC[DEInf$LFC=="-Inf"]<-negINF
DEInf$LFC[DEInf$LFC=="Inf"]<-posINF

columnIfc <- DEInf$FC[which(!is.na(DEInf$FC))]
columnifc <- columnIfc[which(columnIfc < Inf)]
posINFfc<-max(columnifc, na.rm = T)
columnifc <- columnIfc[which(columnIfc > -Inf)]
negINFfc<-min(columnifc, na.rm = T)
DEInf$FC[DEInf$FC=="Inf"]<-posINFfc
####pval->fdr may 21
RepINF<-data.frame(DEInf$Read.Count.All, DEInf$LFC, DEInf$FDR)
colnames(RepINF)<-c("baseMean", "log2FoldChange", "padj")

maPlot<-ggmaplot(RepINF, genenames = NULL, top = 0, palette = c("#B31B21", "#1465AC", "darkgray"),fdr = 0.05, fc = 2, size=0.4, ggtheme = ggplot2::theme_minimal(), ylab = comparision, font.legend =c("bold", 16))+theme(axis.text=element_text(size=16), axis.title =element_text(size=18))

ggsave(maPlot, file=paste0(i,"_MA_images.png"), width = 13, height = 4, path=DEOut)
}
```

```{r DEHeatmaps, echo=FALSE, warning=FALSE, fig.show='hide', message = FALSE}
myinfo <- data.frame(read.table(infofile, check.names=FALSE, sep="\t", as.is = TRUE), check.names=FALSE)
comparision_list<-list()
#normal<-paste0(GEPathD, "logCPM_count_normalized.txt")
DEcounts<-list.files(DECountPath, pattern = ".counts$", full.names = TRUE)
for(i in 1: length(DeFiles))
{
  DEInf<-read.table(DeFiles[i], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  forprop<-DEInf[order(DEInf$Read.Count.All), ]
  DECnts<-read.table(DEcounts[i], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  names(DEInf)[6]<-"LFC"
  DEInf<-DEInf[DEInf$FDR<=FDRc, ]
  DEInf<-DEInf[DEInf$LFC >=LFCc | DEInf$LFC<= -LFCc, ]
  DEInf<-DEInf[order(DEInf$Read.Count.All), ]
  
  
  ###Do I need to change this block? commenting for now May 6 2021.
  p10comp<-(nrow(forprop)*10)/100
  p10compdf<-DEInf[0:p10comp,]
  if(nrow(p10compdf)>1){
    cutoff<-mean(range(p10compdf$Read.Count.All))
    #cutoff<-mean(range(DEInf$Read.Count.All))
  }else {
    cutoff<-mean(range(DEInf$Read.Count.All))
  }
  cutoffdf=forprop[p10comp,]
  cutoff=cutoffdf$Read.Count.All
  DEInf<-DEInf[DEInf$Read.Count.All >= cutoff, ]
  DEnoInf<-DEInf
  
  if(nrow(DEnoInf)>=2)
  {
  columnI <- DEnoInf$LFC[which(!is.na(DEnoInf$LFC))]
  columni <- columnI[which(columnI < Inf)]
  posINF<-max(columni,na.rm = T)
  columni <- columnI[which(columnI > -Inf)]
  negINF<-min(columni,na.rm = T)
  DEnoInf$LFC[DEnoInf$LFC=="-Inf"]<-negINF
  DEnoInf$LFC[DEnoInf$LFC=="Inf"]<-posINF
  Hmap_LFC<-data.frame(DEnoInf$Feature.ID, DEnoInf$LFC)
  Hmap_LFC<-Hmap_LFC[order(Hmap_LFC$DEnoInf.LFC, decreasing = TRUE), ]
  t1_comp<-names(DEnoInf)[3]
  t2_comp<-names(DEnoInf)[4]
  comp_list<-list()
  t1_comp<-gsub("Read.Count.", "", t1_comp)
  t2_comp<-gsub("Read.Count.", "", t2_comp)
  comp_list<-append(comp_list, t1_comp)
  comp_list<-append(comp_list, t2_comp)
  comp_list<-unlist(comp_list)
  degenes_list<-DEnoInf$Feature.ID #DEInf used in normal=T
  comp_samples<-data.frame(subset(myinfo, myinfo$V2 %in%comp_list), check.names=FALSE)
  comp_sampleList<- unlist(comp_samples$V1)
  ##Same Till here
  
  #DECnts_samples<-data.frame(subset(DECnts, colnames(DECnts[2:21]) %in% comp_sampleList), check.names=F)
  DECnts_samples<-DECnts
  rownames(DECnts_samples)<-DECnts_samples$ID
  DECnts_samples$ID<-NULL
  #Remove all rowsums with 0
  DECnts_no0<-data.frame(DECnts_samples[rowSums(DECnts_samples)!=0, ], check.names = FALSE)
  #Replace 0 with 1
  DECnts_no0[DECnts_no0==0]<-1
  DECnts_samples<-DECnts_no0
  Filtered_norm<-subset(DECnts_samples, rownames(DECnts_samples) %in% degenes_list, check.names=FALSE)
  if(length(Filtered_norm)>=2){
  comp_sampleList<-comp_samples[order(comp_samples$V1),]
  heatmp_a<-data.frame(factor(comp_sampleList$V2))
  colass<-unlist(unique(heatmp_a)) #Remove- maybe Unnecessary
  colnames(heatmp_a)<-"Group"   
  levels(heatmp_a)=levels(heatmp_a$Group)
  #levels(heatmp_a)<-heatmp_a
  l1<-levels(heatmp_a$Group)[1]
  l2<-levels(heatmp_a$Group)[2]
  df=data.frame(heatmp_a=heatmp_a)      
  string_command<-paste0('HeatmapAnnotation(df=df, annotation_legend_param = list(title="Condition", title_gp = gpar(fontsize = 10, fontface=2, fontfamily="Helvetica")),col=list(Group=c(',l1,'="darkgreen",', l2, '="blue")))')
  heatmap_a<-eval(parse(text=string_command))
  maxRD<- max(Hmap_LFC$DEnoInf.LFC, na.rm = T)
  minRD<-min(Hmap_LFC$DEnoInf.LFC, na.rm= T)
  Hmap_LFC<-Hmap_LFC[which(!is.na(Hmap_LFC$DEnoInf.Feature.ID)), ]
  rownames(Hmap_LFC)<-Hmap_LFC$DEnoInf.Feature.ID
  resort<-Filtered_norm[match(rownames(Hmap_LFC), rownames(Filtered_norm)),]
  resort<-resort[,order(match(paste(colnames(resort)), paste(comp_sampleList$V1)))]
  haDE.h<-Heatmap(data.frame(Hmap_LFC$DEnoInf.LFC), cluster_rows = FALSE, width = 0.25, show_heatmap_legend = FALSE, show_row_names = FALSE, show_column_names = FALSE, col = colorRamp2(c(minRD, maxRD), c("gray", "pink")))
  hmap_plot<- Heatmap(resort, show_row_names = FALSE, column_names_gp = gpar(fontsize = 6), name="ScaledCPM", column_title_gp = gpar(fontsize=6), column_title_side = "bottom", top_annotation = heatmap_a, row_names_gp = gpar(fontsize = 2), cluster_rows = FALSE,row_title=paste0("Size:",nrow(resort)," RC:",cutoff))
  #print(paste0("Size:",nrow(resort)))
  htlistDE=hmap_plot +  haDE.h
  file_name = paste(DEOut,i, "_Hmap_images.png", sep="")
  png(file_name,width = 7, height = 8, units = 'in', res = 600)
  #plot_list[i] = htlistDE
  draw(htlistDE)
  dev.off()
  }
  }
}
```

```{r imagesplots, echo=FALSE, warning=FALSE, fig.show='hide', message = FALSE, include=FALSE}
Heat_images<-list.files(DEOut, pattern = "_Hmap_images.png", full.names = TRUE)
MA_images <- list.files(DEOut, pattern = "_MA_images.png", full.names = TRUE)
```
\newpage
\begin{center}

Figure 1. DE Heatmaps: This set of figures illustrate the clustering for each
comparison of interest based on the Gene Expression Counts and Log Fold Change
(LFC) for the Differentially Expressed Genes (DEGs). Columns correspond to
samples and rows correspond to genes. Scale bar represents a graded change in LFC where pink is higher LFC, gray is lower LFC. Gene expression valuescome from  DESeq/DESeq2 normalized counts.

\end{center}
```{r Hmapdisp, echo=FALSE, warning=FALSE, message = FALSE, fig.height=8, fig.width=12}
for (i in Heat_images) {
  temp_hmap<-readPNG(i)
  hmap_grob<-grid::rasterGrob(as.raster(temp_hmap))
  grid.arrange(hmap_grob)
}
#include_graphics(Heat_images, dpi=250)
```
\newpage
\begin{center}
Figure 2. MA Plots: This set of plots displays the relationship between
expression change (LFC, M) and average expression (log2 mean expression, A).
Points colored in red indicate the up-regulated genes with FDR <= 0.05 and LFC
>= 1. Points colored in blue indicate down-regulated genes with FDR <= 0.05 and
log2-fold-change >= +/1. Dashed lines indicate the LFC cutoff values of +/ 1.
Points colored in gray indicate those gees that are not significantly
differentially expressed.

\end{center}
\  
```{r MAdisp, echo=FALSE, warning=FALSE, message = FALSE}
#Heat_MA<-list.files(DEOut,pattern = "_DE.png", full.names = TRUE)
include_graphics(MA_images, dpi=600)
```
\newpage
 
```{r Venn, echo=FALSE, warning=FALSE, fig.show='hide'}

comp<-length(DeFiles)
if(comp >=3)
{
  randC<-sample.int(comp, 3)
  comp1<-read.table(DeFiles[randC[1]], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  names(comp1)[6]<-"LFC"
  testcomp1<-comp1[comp1$FDR<FDRc, ]
  Fcomp1<-testcomp1[testcomp1$LFC >LFCc | testcomp1$LFC< -LFCc, ]
  sortcomp1<-comp1[order(comp1$Read.Count.All), ]
  p10comp1<-(nrow(sortcomp1)*10)/100
  p10comp1df<-sortcomp1[0:p10comp1,]
  cutoff1<-mean(range(p10comp1df$Read.Count.All))
  Rcomp1<-Fcomp1[Fcomp1$Read.Count.All > cutoff1, ]
  
  name1<-basename(DeFiles[randC[1]])
  name1<-gsub("\\..*","",name1)
  comp2<-read.table(DeFiles[randC[2]], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  names(comp2)[6]<-"LFC"
  testcomp2<-comp2[comp2$FDR<FDRc, ]
  Fcomp2<-testcomp2[testcomp2$LFC >LFCc | testcomp2$LFC< -LFCc, ]
  name2<-basename(DeFiles[randC[2]])
  name2<-gsub("\\..*","",name2)
  sortcomp2<-comp2[order(comp2$Read.Count.All), ]
  p10comp2<-(nrow(sortcomp2)*10)/100
  p10comp2df<-sortcomp2[0:p10comp2,]
  cutoff2<-mean(range(p10comp2df$Read.Count.All))
  Rcomp2<-Fcomp2[Fcomp2$Read.Count.All > cutoff2, ]
  
  comp3<-read.table(DeFiles[randC[3]], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  names(comp3)[6]<-"LFC"
  testcomp3<-comp3[comp3$FDR<FDRc, ]
  Fcomp3<-testcomp3[testcomp3$LFC >LFCc | testcomp3$LFC< -LFCc, ]
  name3<-basename(DeFiles[randC[3]])
  name3<-gsub("\\..*","",name3)
  sortcomp3<-comp3[order(comp3$Read.Count.All), ]
  p10comp3<-(nrow(sortcomp3)*10)/100
  p10comp3df<-sortcomp3[0:p10comp3,]
  cutoff3<-mean(range(p10comp3df$Read.Count.All))
  Rcomp3<-Fcomp3[Fcomp3$Read.Count.All > cutoff3, ]


mergeComp<-list(Rcomp1$Feature.ID,Rcomp2$Feature.ID, Rcomp3$Feature.ID)
names(mergeComp)<-c(name1, name2, name3)
venn<-venn.diagram(mergeComp, filename = NULL, fill = c("orange", "blue", "green"), cex=1, cat.cex=0.7)
}

if (comp ==2)
{
  comp1<-read.table(DeFiles[1], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  names(comp1)[6]<-"LFC"
  testcomp1<-comp1[comp1$FDR<=FDRc, ]
  Fcomp1<-testcomp1[testcomp1$LFC >LFCc | testcomp1$LFC< -LFCc, ]
  sortcomp1<-comp1[order(comp1$Read.Count.All), ]
  p10comp1<-(nrow(sortcomp1)*10)/100
  p10comp1df<-sortcomp1[0:p10comp1,]
  cutoff1<-mean(range(p10comp1df$Read.Count.All))
  Rcomp1<-Fcomp1[Fcomp1$Read.Count.All > cutoff1, ]
  
  name1<-basename(DeFiles[1])
  name1<-gsub("\\..*","",name1)
  
  comp2<-read.table(DeFiles[2], check.names=FALSE, sep="\t", as.is = TRUE, header = T)
  names(comp2)[6]<-"LFC"
  testcomp2<-comp2[comp2$FDR<FDRc, ]
  Fcomp2<-testcomp2[testcomp2$LFC >LFCc | testcomp2$LFC< -LFCc, ]
  name2<-basename(DeFiles[2])
  name2<-gsub("\\..*","",name2)
  sortcomp2<-comp2[order(comp2$Read.Count.All), ]
  p10comp2<-(nrow(sortcomp2)*10)/100
  p10comp2df<-sortcomp2[0:p10comp2,]
  cutoff2<-mean(range(p10comp2df$Read.Count.All))
  Rcomp2<-Fcomp2[Fcomp2$Read.Count.All > cutoff2, ]
  
  mergeComp<-list(Rcomp1$Feature.ID,Rcomp2$Feature.ID)
names(mergeComp)<-c(name1, name2)
venn<-venn.diagram(mergeComp, filename = NULL, fill = c("orange", "blue"), cex=1, cat.cex=0.7)
  
}
``` 

```{r venndrw, echo=FALSE, warning=FALSE, message=FALSE}
if(comp>1)
{
#grid.draw(venn)
venn_title<-paste0("Figure 3. Venn Diagram: This diagram shows the number of shared DE genes between comparisons ", toString(names(mergeComp)), ". If there are more than three pairwise comparisons available, the comparisons used to show the overlap have been randomly selected.")
venn_wrap<-strwrap(venn_title, width=98, simplify=F)
venn_newline <- sapply(venn_wrap, paste, collapse = "\n")
venn_save<-grid.arrange(gTree(children=venn), top=textGrob(venn_newline, gp=gpar(fontsize=9)))
ggsave(paste0(DEOut,"/","VennDiagram",".png"), venn_save)
}
```
\newpage
```{r scatter, echo=FALSE, warning=FALSE, fig.height=6, fig.width=8, message=FALSE}

#ggsave(paste0(DEOut,"/","VennDiagram",".png"), venn_save)

col <- c()
col2 <- c()
col3 <- c()
lab <- c()
lab2<- c()
lab3<- c()

sum_1 <- 0
sum_2 <- 0
sum_3 <- 0
sum_4 <- 0
sum_5 <- 0
sum_6 <- 0
sum_7 <- 0
sum_8 <- 0

if (comp ==2)
{
var1<-data.frame(Rcomp1$Feature.ID, Rcomp1$LFC)
var2<-data.frame(Rcomp2$Feature.ID, Rcomp2$LFC)
colnames(var1)<-c("Feature.ID", "LFC1")
colnames(var2)<-c("Feature.ID", "LFC2")

mergeRcomp<-merge(var1, var2, by.x='Feature.ID', by.y='Feature.ID', all=TRUE)
mergeRcomp[is.na(mergeRcomp)]<- 0
plot_df<-mergeRcomp
plot_df["bin"]<-0

for(i in 1:nrow(mergeRcomp)){ 
 if(mergeRcomp$LFC1[i]<0 && mergeRcomp$LFC2[i]<0)
 {
   plot_df$bin[i]=3
 }
  if(mergeRcomp$LFC1[i]>0 && mergeRcomp$LFC2[i]>0)
 {
   plot_df$bin[i]=1
  }
  if(mergeRcomp$LFC1[i]<0 && mergeRcomp$LFC2[i]>0)
 {
   plot_df$bin[i]=2
  }
  if(mergeRcomp$LFC1[i]>0 && mergeRcomp$LFC2[i]<0)
 {
   plot_df$bin[i]=4
  }
  if(mergeRcomp$LFC1[i]>0 && mergeRcomp$LFC2[i]==0)
 {
   plot_df$bin[i]=5
  }
  if(mergeRcomp$LFC1[i]==0 && mergeRcomp$LFC2[i]>0)
 {
   plot_df$bin[i]=6
  }
  if(mergeRcomp$LFC1[i]<0 && mergeRcomp$LFC2[i]==0)
 {
   plot_df$bin[i]=7
  }
  if(mergeRcomp$LFC1[i]==0 && mergeRcomp$LFC2[i]<0)
 {
   plot_df$bin[i]=8
 }
   
}

sum_1<-0
sum_2<-0
sum_3<-0
sum_4<-0
sum_5<-0
sum_6<-0
sum_7<-0
sum_8<-0

for(i in 1:nrow(plot_df)){ 
		if (plot_df$bin[i] == 1 ){sum_1 <- sum_1 + 1} 
		if (plot_df$bin[i] == 2 ){sum_2 <- sum_2 + 1} 
		if (plot_df$bin[i] == 3 ){sum_3 <- sum_3 + 1} 
		if (plot_df$bin[i] == 4 ){sum_4 <- sum_4 + 1} 
		if (plot_df$bin[i] == 5 ){sum_5 <- sum_5 + 1} 
		if (plot_df$bin[i] == 6 ){sum_6 <- sum_6 + 1} 
		if (plot_df$bin[i] == 7 ){sum_7 <- sum_7 + 1} 
		if (plot_df$bin[i] == 8 ){sum_8 <- sum_8 + 1} 
	}

if (sum_1 >= 1) {
		col <- append(col, "red",)
		lab <- append(lab,paste("UP/UP",":",sum_1,sep='')) 
	}
	if (sum_2 >= 1) {
		col <- append(col, "blue")
		lab <- append(lab,paste("DOWN/UP",":",sum_2,sep='')) 	
	}
	if (sum_3 >= 1) {
		col <- append(col, "black")
		lab <- append(lab,paste("DOWN/DOWN",":",sum_3,sep='')) 	
	}
	if (sum_4 >= 1) {
		col <- append(col, "green")
		lab <- append(lab,paste("UP/DOWN",":",sum_4,sep='')) 	
	}
	if (sum_5 >= 1) {
		col <- append(col, "brown")
		lab <- append(lab,paste("UP/NONE",":",sum_5,sep='')) 	
	}
	if (sum_6 >= 1) {
		col <- append(col, "cyan")
		lab <- append(lab,paste("NONE/UP",":",sum_6,sep='')) 	
	}
	if (sum_7 >= 1) {
		col <- append(col, "dark green")
		lab <- append(lab,paste("DOWN/NONE",":",sum_7,sep='')) 	
	}
	if (sum_8 >= 1) {
		col <- append(col, "purple")
		lab <- append(lab,paste("NONE/DOWN",":",sum_8,sep='')) 	
	}
heading1 <- paste(name1, name2, sep=' and ')
title_qplot1<-paste0("Figure 4. Quadrant Plots: This set of plots show a comparison of DEGs for each of the pairwise comparisons. Each point represents a DEG. Red points in the upper right hand quadrant are genes with increased expression in both comparisons. Black points in the bottom left hand quadrant are genes with decreased expression in both comparisons. The points in the other two quadrants show genes that have opposite regulation between the two comparisons.")
q_wrap<-strwrap(title_qplot1, width=80, simplify=F)
q_newline <- sapply(q_wrap, paste, collapse = "\n")
hold<-ggplot(data=plot_df, aes(x=LFC1 , y=LFC2, colour=as.character(bin))) + theme_bw() + geom_point(alpha=1,size=1) + xlab(paste(name1,"(logFoldChange)",sep='')) + ylab(paste(name2,"(logFoldChange)",sep='')) + scale_colour_manual(name="Log(Fold.Change)",values=col,labels=lab)
qplot_s<-grid.arrange(hold, top=textGrob(q_newline, gp=gpar(fontsize=11)))
ggsave(paste0(DEOut,"/",heading1,".png"), qplot_s)
}

if (comp >=3)
{
var1<-data.frame(Rcomp1$Feature.ID, Rcomp1$LFC)
var2<-data.frame(Rcomp2$Feature.ID, Rcomp2$LFC)
var3<-data.frame(Rcomp3$Feature.ID, Rcomp3$LFC)
colnames(var1)<-c("Feature.ID", "LFC1")
colnames(var2)<-c("Feature.ID", "LFC2")
colnames(var3)<-c("Feature.ID", "LFC3")

mergeRcomp<-merge(var1, var2, by.x='Feature.ID', by.y='Feature.ID', all=TRUE)
mergeRcomp[is.na(mergeRcomp)]<- 0
plot_df<-mergeRcomp
plot_df["bin"]<-0


for(i in 1:nrow(mergeRcomp)){ 
 if(mergeRcomp$LFC1[i]<0 && mergeRcomp$LFC2[i]<0)
 {
   plot_df$bin[i]=3
 }
  if(mergeRcomp$LFC1[i]>0 && mergeRcomp$LFC2[i]>0)
 {
   plot_df$bin[i]=1
  }
  if(mergeRcomp$LFC1[i]<0 && mergeRcomp$LFC2[i]>0)
 {
   plot_df$bin[i]=2
  }
  if(mergeRcomp$LFC1[i]>0 && mergeRcomp$LFC2[i]<0)
 {
   plot_df$bin[i]=4
  }
  if(mergeRcomp$LFC1[i]>0 && mergeRcomp$LFC2[i]==0)
 {
   plot_df$bin[i]=5
  }
  if(mergeRcomp$LFC1[i]==0 && mergeRcomp$LFC2[i]>0)
 {
   plot_df$bin[i]=6
  }
  if(mergeRcomp$LFC1[i]<0 && mergeRcomp$LFC2[i]==0)
 {
   plot_df$bin[i]=7
  }
  if(mergeRcomp$LFC1[i]==0 && mergeRcomp$LFC2[i]<0)
 {
   plot_df$bin[i]=8
 }
   
}
sum_1<-0
sum_2<-0
sum_3<-0
sum_4<-0
sum_5<-0
sum_6<-0
sum_7<-0
sum_8<-0

for(i in 1:nrow(plot_df)){ 
		if (plot_df$bin[i] == 1 ){sum_1 <- sum_1 + 1} 
		if (plot_df$bin[i] == 2 ){sum_2 <- sum_2 + 1} 
		if (plot_df$bin[i] == 3 ){sum_3 <- sum_3 + 1} 
		if (plot_df$bin[i] == 4 ){sum_4 <- sum_4 + 1} 
		if (plot_df$bin[i] == 5 ){sum_5 <- sum_5 + 1} 
		if (plot_df$bin[i] == 6 ){sum_6 <- sum_6 + 1} 
		if (plot_df$bin[i] == 7 ){sum_7 <- sum_7 + 1} 
		if (plot_df$bin[i] == 8 ){sum_8 <- sum_8 + 1} 
	}

if (sum_1 >= 1) {
		col <- append(col, "red",)
		lab <- append(lab,paste("UP/UP",":",sum_1,sep='')) 
	}
	if (sum_2 >= 1) {
		col <- append(col, "blue")
		lab <- append(lab,paste("DOWN/UP",":",sum_2,sep='')) 	
	}
	if (sum_3 >= 1) {
		col <- append(col, "black")
		lab <- append(lab,paste("DOWN/DOWN",":",sum_3,sep='')) 	
	}
	if (sum_4 >= 1) {
		col <- append(col, "green")
		lab <- append(lab,paste("UP/DOWN",":",sum_4,sep='')) 	
	}
	if (sum_5 >= 1) {
		col <- append(col, "brown")
		lab <- append(lab,paste("UP/NONE",":",sum_5,sep='')) 	
	}
	if (sum_6 >= 1) {
		col <- append(col, "cyan")
		lab <- append(lab,paste("NONE/UP",":",sum_6,sep='')) 	
	}
	if (sum_7 >= 1) {
		col <- append(col, "dark green")
		lab <- append(lab,paste("DOWN/NONE",":",sum_7,sep='')) 	
	}
	if (sum_8 >= 1) {
		col <- append(col, "purple")
		lab <- append(lab,paste("NONE/DOWN",":",sum_8,sep='')) 	
	}
#heading1 <- paste(name1, name2, sep='_')
heading1 <- paste(name1, name2, sep=' and ')
#title_qplot1<-paste("Figure 4. Quadrant plot showing the consistency in the up- and down-regulation of in differential gene expression between comparisons.", heading1)
title_qplot1<-paste0("Figure 4. Quadrant plot showing a comparison of DEGs for each of the pairwise comparisons. Each point represents a DEG. Red points in the upper right hand quadrant are genes with increased expression in both comparisons. Black points in the bottom left hand quadrant are genes with decreased expression in both comparisons. The points in the other two quadrants show genes that have opposite regulation between the two comparisons.")
q_wrap<-strwrap(title_qplot1, width=98, simplify=F)
q_newline <- sapply(q_wrap, paste, collapse = "\n")
Qplot1<-ggplot(data=plot_df, aes(x=LFC1 , y=LFC2, colour=as.character(bin))) + theme_bw() + geom_vline(xintercept = 0) + geom_hline(yintercept = 0) + geom_point(alpha=1,size=1) + xlab(paste(name1,"(logFoldChange)",sep='')) + ylab(paste(name2,"(logFoldChange)",sep='')) + scale_colour_manual(name="Log(Fold.Change)",values=col,labels=lab)
ggsave(paste0(DEOut,"/",heading1,".png"), Qplot1)
####Plot2

mergeRcomp2<-merge(var1, var3, by.x='Feature.ID', by.y='Feature.ID', all=TRUE)
mergeRcomp2[is.na(mergeRcomp2)]<- 0
plot_df2<-mergeRcomp2
plot_df2["bin"]<-0

for(i in 1:nrow(mergeRcomp2)){ 
 if(mergeRcomp2$LFC1[i]<0 && mergeRcomp2$LFC3[i]<0)
 {
   plot_df2$bin[i]=3
 }
  if(mergeRcomp2$LFC1[i]>0 && mergeRcomp2$LFC3[i]>0)
 {
   plot_df2$bin[i]=1
  }
  if(mergeRcomp2$LFC1[i]<0 && mergeRcomp2$LFC3[i]>0)
 {
   plot_df2$bin[i]=2
  }
  if(mergeRcomp2$LFC1[i]>0 && mergeRcomp2$LFC3[i]<0)
 {
   plot_df2$bin[i]=4
  }
  if(mergeRcomp2$LFC1[i]>0 && mergeRcomp2$LFC3[i]==0)
 {
   plot_df2$bin[i]=5
  }
  if(mergeRcomp2$LFC1[i]==0 && mergeRcomp2$LFC3[i]>0)
 {
   plot_df2$bin[i]=6
  }
  if(mergeRcomp2$LFC1[i]<0 && mergeRcomp2$LFC3[i]==0)
 {
   plot_df2$bin[i]=7
  }
  if(mergeRcomp2$LFC1[i]==0 && mergeRcomp2$LFC3[i]<0)
 {
   plot_df2$bin[i]=8
 }
   
}
sum_1<-0
sum_2<-0
sum_3<-0
sum_4<-0
sum_5<-0
sum_6<-0
sum_7<-0
sum_8<-0

for(i in 1:nrow(plot_df2)){ 
		if (plot_df2$bin[i] == 1 ){sum_1 <- sum_1 + 1} 
		if (plot_df2$bin[i] == 2 ){sum_2 <- sum_2 + 1} 
		if (plot_df2$bin[i] == 3 ){sum_3 <- sum_3 + 1} 
		if (plot_df2$bin[i] == 4 ){sum_4 <- sum_4 + 1} 
		if (plot_df2$bin[i] == 5 ){sum_5 <- sum_5 + 1} 
		if (plot_df2$bin[i] == 6 ){sum_6 <- sum_6 + 1} 
		if (plot_df2$bin[i] == 7 ){sum_7 <- sum_7 + 1} 
		if (plot_df2$bin[i] == 8 ){sum_8 <- sum_8 + 1} 
	}

if (sum_1 >= 1) {
		col2 <- append(col2, "red",)
		lab2 <- append(lab2,paste("UP/UP",":",sum_1,sep='')) 
	}
	if (sum_2 >= 1) {
		col2 <- append(col2, "blue")
		lab2 <- append(lab2,paste("DOWN/UP",":",sum_2,sep='')) 	
	}
	if (sum_3 >= 1) {
		col2 <- append(col2, "black")
		lab2 <- append(lab2,paste("DOWN/DOWN",":",sum_3,sep='')) 	
	}
	if (sum_4 >= 1) {
		col2 <- append(col2, "green")
		lab2 <- append(lab2,paste("UP/DOWN",":",sum_4,sep='')) 	
	}
	if (sum_5 >= 1) {
		col2 <- append(col2, "brown")
		lab2 <- append(lab2,paste("UP/NONE",":",sum_5,sep='')) 	
	}
	if (sum_6 >= 1) {
		col2 <- append(col2, "cyan")
		lab2 <- append(lab2,paste("NONE/UP",":",sum_6,sep='')) 	
	}
	if (sum_7 >= 1) {
		col2 <- append(col2, "dark green")
		lab2 <- append(lab2,paste("DOWN/NONE",":",sum_7,sep='')) 	
	}
	if (sum_8 >= 1) {
		col2 <- append(col2, "purple")
		lab2 <- append(lab2,paste("NONE/DOWN",":",sum_8,sep='')) 	
	}
heading2 <- paste(name1, name3, sep=' and ')
title_qplot2<-paste("Quadrant Plot showing differentially expressed genes between", heading2)
#Qplot2<-ggplot(data=plot_df2, aes(x=LFC1 , y=LFC3, colour=as.character(bin))) + theme_bw() + geom_vline(xintercept = 0) + geom_hline(yintercept = 0)+ geom_point(alpha=1,size=1) + xlab(paste(name1,"(logFoldChange)",sep='')) + ylab(paste(name3,"(logFoldChange)",sep=''))+ labs(title=heading2) + scale_colour_manual(name="Log(Fold.Change)",values=col2,labels=lab2)

Qplot2<-ggplot(data=plot_df2, aes(x=LFC1 , y=LFC3, colour=as.character(bin))) + theme_bw() + geom_vline(xintercept = 0) + geom_hline(yintercept = 0)+ geom_point(alpha=1,size=1) + xlab(paste(name1,"(logFoldChange)",sep='')) + ylab(paste(name3,"(logFoldChange)",sep='')) + scale_colour_manual(name="Log(Fold.Change)",values=col2,labels=lab2)

ggsave(paste0(DEOut,"/",heading2,".png"), Qplot2)
mergeRcomp3<-merge(var2, var3, by.x='Feature.ID', by.y='Feature.ID', all=TRUE)
mergeRcomp3[is.na(mergeRcomp3)]<- 0
plot_df3<-mergeRcomp3
plot_df3["bin"]<-0

for(i in 1:nrow(mergeRcomp3)){ 
 if(mergeRcomp3$LFC2[i]<0 && mergeRcomp3$LFC3[i]<0)
 {
   plot_df3$bin[i]=3
 }
  if(mergeRcomp3$LFC2[i]>0 && mergeRcomp3$LFC3[i]>0)
 {
   plot_df3$bin[i]=1
  }
  if(mergeRcomp3$LFC2[i]<0 && mergeRcomp3$LFC3[i]>0)
 {
   plot_df3$bin[i]=2
  }
  if(mergeRcomp3$LFC2[i]>0 && mergeRcomp3$LFC3[i]<0)
 {
   plot_df3$bin[i]=4
  }
  if(mergeRcomp3$LFC2[i]>0 && mergeRcomp3$LFC3[i]==0)
 {
   plot_df3$bin[i]=5
  }
  if(mergeRcomp3$LFC2[i]==0 && mergeRcomp3$LFC3[i]>0)
 {
   plot_df3$bin[i]=6
  }
  if(mergeRcomp3$LFC2[i]<0 && mergeRcomp3$LFC3[i]==0)
 {
   plot_df3$bin[i]=7
  }
  if(mergeRcomp3$LFC2[i]==0 && mergeRcomp3$LFC3[i]<0)
 {
   plot_df3$bin[i]=8
 }
   
}
sum_1<-0
sum_2<-0
sum_3<-0
sum_4<-0
sum_5<-0
sum_6<-0
sum_7<-0
sum_8<-0
for(i in 1:nrow(plot_df3)){ 
		if (plot_df3$bin[i] == 1 ){sum_1 <- sum_1 + 1} 
		if (plot_df3$bin[i] == 2 ){sum_2 <- sum_2 + 1} 
		if (plot_df3$bin[i] == 3 ){sum_3 <- sum_3 + 1} 
		if (plot_df3$bin[i] == 4 ){sum_4 <- sum_4 + 1} 
		if (plot_df3$bin[i] == 5 ){sum_5 <- sum_5 + 1} 
		if (plot_df3$bin[i] == 6 ){sum_6 <- sum_6 + 1} 
		if (plot_df3$bin[i] == 7 ){sum_7 <- sum_7 + 1} 
		if (plot_df3$bin[i] == 8 ){sum_8 <- sum_8 + 1} 
	}

if (sum_1 >= 1) {
		col3 <- append(col3, "red",)
		lab3 <- append(lab3,paste("UP/UP",":",sum_1,sep='')) 
	}
	if (sum_2 >= 1) {
		col3 <- append(col3, "blue")
		lab3 <- append(lab3,paste("DOWN/UP",":",sum_2,sep='')) 	
	}
	if (sum_3 >= 1) {
		col3 <- append(col3, "black")
		lab3 <- append(lab3,paste("DOWN/DOWN",":",sum_3,sep='')) 	
	}
	if (sum_4 >= 1) {
		col3 <- append(col3, "green")
		lab3 <- append(lab3,paste("UP/DOWN",":",sum_4,sep='')) 	
	}
	if (sum_5 >= 1) {
		col3 <- append(col3, "brown")
		lab3 <- append(lab3,paste("UP/NONE",":",sum_5,sep='')) 	
	}
	if (sum_6 >= 1) {
		col3 <- append(col3, "cyan")
		lab3 <- append(lab3,paste("NONE/UP",":",sum_6,sep='')) 	
	}
	if (sum_7 >= 1) {
		col3 <- append(col3, "dark green")
		lab3 <- append(lab3,paste("DOWN/NONE",":",sum_7,sep='')) 	
	}
	if (sum_8 >= 1) {
		col3 <- append(col3, "purple")
		lab3 <- append(lab3,paste("NONE/DOWN",":",sum_8,sep='')) 	
	}

heading3 <- paste(name2, name3, sep=' and ')
title_qplot3<-paste("Quadrant Plot showing differentially expressed genes between", heading3)
Qplot3<- ggplot(data=plot_df3, aes(x=LFC2 , y=LFC3, colour=as.character(bin))) + theme_bw() + geom_vline(xintercept = 0) + geom_hline(yintercept = 0)+ geom_point(alpha=1,size=1) + xlab(paste(name2,"(logFoldChange)",sep='')) + ylab(paste(name3,"(logFoldChange)",sep='')) + scale_colour_manual(name="Log(Fold.Change)",values=col3,labels=lab3)
ggsave(paste0(DEOut,"/",heading3,".png"), Qplot3)
grid.arrange(Qplot1, top=textGrob(q_newline, gp=gpar(fontsize=11)))
grid.arrange(Qplot2)
grid.arrange(Qplot3)
}

```
